docs/developer-docs/security/rust-canister-development-security-best-practices.md

```diff
@@ -461,6 +461,99 @@ Using the rust CDK, the reccuring timer is also lost on upgrade as explained in
+
+You should consider such rate limits when designing and implementing your canisters. Rate limits are enforced using different time granularities, e.g., seconds or minutes. For second-granularity enforcement, make sure that the simultaneous requests by all subnet replicas do not violate the quota. Violations may lead to temporary or permanent bans.
+
+[More information.](https://internetcomputer.org/docs/current/developer-docs/integrations/https-outcalls/https-outcalls-how-it-works#rate-limiting-by-servers)
+
+### Only make HTTP outcall requests to idempotent endpoints
+
+#### Security concern
+
+As mentioned before, if an HTTP outcall is performed it is amplified by the number of replicas in the subnet. That means the queried endpoint will receive the same request several times. This is especially risky in requests that change the endpoint state given one HTTP outcall could lead to unintentionally changing the endpoint state several times.
+
+#### Recommendation
+
+Make sure the endpoints, called by an HTTP outcall, are idempotent, i.e. the queried endpoint has the same behaviour with the same request payload, no matter the number of times it is called.
+
+Some servers support the use of idempotency keys. These keys are random unique strings submitted in the HTTP request as headers. If used with the HTTP outcalls feature, all requests sent by each (honest) replica will contain the same idempotency key. This allows the server to recognize duplicated requests (i.e requests with the same idempotency key), handle just one and modify the server state only once. Note that this is a feature that must be supported by the server.
+
+[More information.](https://internetcomputer.org/docs/current/developer-docs/integrations/https-outcalls/https-outcalls-how-it-works#post-requests-must-be-idempotent)
+
+### Ensure HTTP responses are identical
+
+#### Security concern
+
+When replicas of a subnet receive HTTP responses, these responses must be identical. Otherwise the consensus won’t be achieved and the HTTP response will be rejected, but still charged.
+
+#### Recommendation
+
+Make sure the HTTP responses sent to the consensus layer are identical.
+
+Ideally the HTTP responses returned by the queried endpoint would be always the same. However most of the time this is not possible to control and the responses include random data (e.g the response includes timestamps, cookie values or some sort of identifiers). In those cases make sure to use the [transformation functions](https://internetcomputer.org/docs/current/developer-docs/integrations/https-outcalls/https-outcalls-how-it-works#transformation-function) to guarantee that the responses, received by each replica, are identical by removing any random data or extracting only the relevant data.
+
+This applies to the HTTP response body and header. Make sure to consider both when applying the transformation functions. Response headers are often overlooked and lead to failure because of failed consensus.
```

Comment:
nit: 
```suggestion
This applies to the HTTP response body and headers. Make sure to consider both when applying the transformation functions. Response headers are often overlooked and lead to failure because of failed consensus.
```